db/virtual_tables.cc

```diff
@@ -975,12 +983,185 @@ class raft_state_table : public streaming_virtual_table {
     }
 };
 
+/// Base class for virtual tables which are supposed to serve data from group0 leader.
+/// If queried on a different node, read will be redirected to the loader.
+/// Implementations need to override execute_on_leader(), which will be executed
+/// on shard0 of current group0 leader.
+///
+/// Current implementation is suitable for tables which are relatively small as all
+/// data is materialized in memory and queried in one page.
+class group0_virtual_table : public memtable_filling_virtual_table {
+private:
```

Comment:
Probably, we should move virtual tables close to where the data they serve is generated (but I don't propose doing that now) so we don't have translation units that depend on the entire universe.